TLS creation/ renewal instructions (#172)
astronomer/docs: enterprise/next/03_install/01_aws/01_install-aws-standard.md

> To install Astronomer on EKS, you'll need access to the following tools and permissions:

Author: Are there any version-specific considerations for this setup? Currently, I'm planning on having the exact same text for all Astronomer versions.

Reviewer: Astronomer 0.16 supports Kubernetes 1.14-1.18 and Astronomer 0.23 supports Kubernetes 1.16-1.18

To install Astronomer on EKS, you'll need access to the following tools and permissions:

* The [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html)
* Kubernetes 1.16 or later
* The [Kubernetes CLI (kubectl)](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
* The [OpenSSL CLI](https://www.openssl.org/docs/man1.0.2/man1/openssl.html)
* [Helm v3.2.1](https://github.com/helm/helm/releases/tag/v3.2.1)
* An SMTP Service & Credentials (e.g. Mailgun, Sendgrid, etc.)
* Permission to create and modify resources on AWS
* Permission to generate a certificate (not self-signed) that covers a defined set of subdomains

> ## Step 1: Choose a Base Domain

Reviewer 2: @jwitz I think in other docs we have steps listed as `## Step 1.` (with no `:`). Should we standardize this + add to style guide?

Author: Yep, I added a bit on this to the style guide earlier this week.

$ kubectl create ns astronomer
```

> ## Step 4: Configure TLS

Author: Should we be mentioning SSL at all in either document? I went with TLS because that's how Steven wrote it in his draft docs, but we've traditional used SSL.

Reviewer: tldr: TLS

TLS is the technically correct term, but many people still say SSL because that's what they are used / it was around longer. The name was changed to Transport Layer Security from Secure Socket Layer to align with the naming convention of OSI model layer 4.

$ kubectl create ns astronomer
```

> ## Step 4: Configure TLS

Reviewer 2: @jwitz I love the added detail in here - it sounds like this is a pain point for a lot of users and this detailed information will better guide them.

Do you think given the breadth here that it'd be worth splitting this content into the other doc and just guide people there instead?

prometheus.BASEDOMAIN
```

To obtain a TLS certificate, complete one of the following setup options:

* **Option 1:** Obtain a TLS certificate from Let's Encrypt. We recommend this option for smaller organizations where your DNS administrator and Kubernetes cluster administrator are either the same person or on the same team.

> * **Option 2:** Request a TLS certificate from your organization's security team. We recommend this option for large organizations with their own  protocols for generating TLS certificates.

Reviewer: ```suggestion
* **Option 2:** Request a TLS certificate from your organization's security team. We recommend this option for large organizations with their own protocols for generating TLS certificates.
```

```
If you are on Linux, run the following:
```sh
$ docker run -it --rm --name letsencrypt -v /etc/letsencrypt:/etc/letsencrypt -v /var/lib/letsencrypt:/var/lib/letsencrypt certbot/certbot:latest certonly -d "*.astro.BASEDOMAIN.com" --manual --preferred-challenges dns --server https://acme-v02.api.letsencrypt.org/directory
```

2. Follow the on-screen prompts and create a TXT record through your DNS provider.

3. Create a Kubernetes Secret named `astronomer-tls` that points to your certificate:
```sh
$ sudo kubectl create secret tls astronomer-tls --key /etc/letsencrypt/live/astro.BASEDOMAIN.com/privkey.pem --cert /etc/letsencrypt/live/astro.BASEDOMAIN.com/fullchain.pem --namespace <your-namespace>
```

> **Note:** If you'd like to use another TLS certificate authority, replace the paths to the Let's Encrypt cert and key .pem files with the paths to your certification's files in the command above.
> ```bash
> kubectl create secret tls astronomer-tls --key <your-private-key-filepath> --cert <your-certificate-filepath> --namespace <your-namespace>
> ```

### Option 2: Request a TLS certificate from your security team

> If you're installing Astronomer for a large organization, you'll need to request a TLS certificate and private key from your enterprise security team. This certificate needs to be valid for your organization's `BASEDOMAIN`, as well as the subdomains listed at the beginning of Step 4. You should be given two `.pem` files: one for your encrypted certificate and one for your private key.

Reviewer: ```suggestion
If you're installing Astronomer for a large organization, you'll need to request a TLS certificate and private key from your enterprise security team. This certificate needs to be valid for the `BASEDOMAIN` selected to be used for Astronomer, as well as the subdomains listed at the beginning of Step 4. You should be given two `.pem` files: one for your encrypted certificate and one for your private key.
```

Reviewer: An important detail is that it does not need to be the organization's own basedomain. This would cause an issue with some restrictive customers. i.e. Example org might not want to issue ...astro.example.com and instead elect to use ....astro.example-astronomer.com.

Reviewer: any domain they own is fine, basically. And the BASEDOMAIN may or may not include subdomains. For example, Astronomer owns the domain astronomer-development.com. When I choose a basedomain, I choose "steven.astronomer-development.com". it would also work to choose "astronomer-development.com" as the basedomain, but that might get in the way if other developers want to use the same domain (e.g. jake.astronomer-development.com).

Author: Ah I see, thanks for mentioning this. I'm updating this wording in my next commit.

To confirm that your enterprise security team generated the correct certificate, run the following command using the openssl command line tool:

```sh
openssl x509 -in  <your-certificate-filepath> -text -noout
```

This command will generate a report. If the `X509v3 Subject Alternative Name` section of this report includes either a single `*.BASEDOMAIN` wildcard domain or the subdomains listed at the beginning of Step 4, then the certificate creation was successful.

Depending on your organization, you might receive either a globally trusted certificate or a certificate from a private certificate authority. If you received a globally trusted certificate, simply run the following command and proceed to Step 5:

```sh
kubectl create secret tls astronomer-tls --cert <your-certificate-filepath> --key <your-private-key-filepath>
```

> If you received a certificate from a private certificate authority, complete the following setup instead:
> 
> 1. Generate Astronomer certificates using the following command:
> ```sh
> openssl x509 -req -in astronomer.csr -CA privateCA.pem -CAkey privateCA.key -CAcreateserial \ -out astronomer.crt -days 825 -sha256 -extfile astronomer.ext
> ```
> 2. Add the root certificate provided by your security team to a Kubernetes secret in the Astronomer namespace using the following command:
> ```sh
> kubectl create secret generic private-root-ca --from-file=cert.pem=./<your-certificate-filepath>
> ```

Reviewer: ```suggestion
If you received a certificate from a private certificate authority, also do the following:

1. Add the root certificate provided by your security team to a Kubernetes secret in the Astronomer namespace using the following command. This should be the certificate of the authority that signed the Astronomer certificate, but not the Astronomer certificate itself. This is the same certificate you typically need to install in clients to get them to trust your services:
```sh
kubectl create secret generic private-root-ca --from-file=cert.pem=./<your-certificate-filepath>
```
```

Reviewer: Feel free to readjust this, but step 1 is performed by the enterprise security team, not those obtaining the certificate.

prometheus.BASEDOMAIN
```

To obtain a TLS certificate, you can:

* **Option 1:** Obtain a TLS certificate from Let's Encrypt. We recommend this option for smaller organizations where your DNS administrator and Kubernetes cluster administrator are either the same person or on the same team.
* **Option 2:** Request a TLS certificate from your organization's security team. We recommend this option for large organizations with their own  protocols for generating TLS certificates.

Complete one of the two setups below based on your organization's needs.

### Option 1: Obtain a TLS certificate from Let's Encrypt

Let's Encrypt is a free and secure service that provides automated TLS certificates. Use this option if you are configuring Astronomer for a smaller organization without a dedicated security team.

1. Obtain a Free SSL Certificate from Let's Encrypt. If you are on a Mac, run the following:

> ```sh

Author: This setup was not in Steven's docs for getting a certificate - wasn't sure whether this is an alternative to/in addition to the setup he described.

Reviewer: This way gets you certificate from Let's Encrypt one time. The cert manager piece is there to perform this procedure on the user's behalf and automatically before the certificate expires.

Author: Sounds good - I'll just link out to the automatic renewal setup in the new doc and remove this setup.

prometheus.BASEDOMAIN
```

To obtain a TLS certificate, complete one of the following setup options:

* **Option 1:** Obtain a TLS certificate from Let's Encrypt. We recommend this option for smaller organizations where your DNS administrator and Kubernetes cluster administrator are either the same person or on the same team.
* **Option 2:** Request a TLS certificate from your organization's security team. We recommend this option for large organizations with their own  protocols for generating TLS certificates.

> ### Option 1: Create automatically renewed TLS certificates using Let's Encrypt

Reviewer 2: @jwitz A bit confused around this header. Does "automatically renewed" mean to describe the certificate? Would it be better to say, `Create TLS certificates that automatically renew using Let's Encrypt`?

If you're installing Astronomer for a large organization, you'll need to request a TLS certificate and private key from your enterprise security team. This certificate needs to be valid for the `BASEDOMAIN` your organization uses for Astronomer, as well as the subdomains listed at the beginning of Step 4. You should be given two `.pem` files: one for your encrypted certificate and one for your private key.

To confirm that your enterprise security team generated the correct certificate, run the following command using the openssl command line tool:

```sh
openssl x509 -in  <your-certificate-filepath> -text -noout
```

This command will generate a report. If the `X509v3 Subject Alternative Name` section of this report includes either a single `*.BASEDOMAIN` wildcard domain or the subdomains listed at the beginning of Step 4, then the certificate creation was successful.

Depending on your organization, you might receive either a globally trusted certificate or a certificate from a private certificate authority. If you received a globally trusted certificate, simply run the following command and proceed to Step 5:

```sh

> kubectl create secret tls astronomer-tls --cert <your-certificate-filepath> --key <your-private-key-filepath>

Reviewer 2: ```suggestion
$ kubectl create secret tls astronomer-tls --cert <your-certificate-filepath> --key <your-private-key-filepath>
```

```

If you received a certificate from a private certificate authority, complete the following setup instead:

1. Add the root certificate provided by your security team to a Kubernetes secret in the Astronomer namespace using the following command:
```sh

> kubectl create secret generic private-root-ca --from-file=cert.pem=./<your-certificate-filepath>

Reviewer 2: ```suggestion
$ kubectl create secret generic private-root-ca --from-file=cert.pem=./<your-certificate-filepath>
```

In the newly created file, copy the example below and replace `baseDomain`, `private-root-ca`, `/etc/docker/certs.d`, `ssl.enabled`, and `smtpUrl` with your own values. For more example configuration files, go [here](https://github.com/astronomer/astronomer/tree/master/configs).

```yaml
#################################
### Astronomer global configuration
#################################
global:
  # Base domain for all subdomains exposed through ingress
  baseDomain: astro.mydomain.com

  # Name of secret containing TLS certificate
  tlsSecret: astronomer-tls

  # Enable privateCaCerts only if your enterprise security team
  # generated a certificate from a private certificate authority.
  privateCaCerts:
  - private-root-ca

  # Enable privateCaCertsAddToHost only when your nodes do not already
  # include the private CA in their docker trust store.
  # Most enterprises already have this configured,
  # and in that case 'enabled' should be false.
  privateCaCertsAddToHost:
    enabled: true
    hostDirectory: /etc/docker/certs.d
  # For development or proof-of-concept, you can use an in-cluster database
  postgresqlEnabled: false

> 

Author: @ianstanton Added this to the config.yaml for AWS, and if it looks good here then I'll copy if over to GCP as well.

   # SSL support for using SSL connections to
   # encrypt client/server communication
   # between database and the Astronomer platform.
   # If your database enforces SSL for connections,
   # change this value to true

>      ssl:
>      enabled: false

Reviewer 3: ```suggestion
  ssl:
    enabled: false
```

Description looks good! Just some indentation fixes.

Author: Sweet! Updated.

### Astronomer global configuration
#################################
global:
  # Base domain for all subdomains exposed through ingress
  baseDomain: astro.mydomain.com

  # Name of secret containing TLS certificate
  tlsSecret: astronomer-tls
  # Only enable privateCaCerts if your enterprise security team
  # generated a certificate from a private certificate authority.
  privateCaCerts:
  - private-root-ca
  # Only enable privateCaCertsAddToHost when your nodes do not already
  # include the private CA in their docker trust store.
  # Most enterprises already have this configured,
  # and in that case 'enabled' should be false.
  privateCaCertsAddToHost:
    enabled: true
    hostDirectory: /etc/docker/certs.d

>   postgresqlEnabled: true

Reviewer: ```suggestion
  # For development or proof of concept only, you can use an in-cluster database
  postgresqlEnabled: false
```

#################################
### Nginx configuration
#################################
nginx:
  # IP address the nginx ingress should bind to
  loadBalancerIP: ~

#################################
### SMTP configuration
#################################

astronomer:
  houston:
    config:
      publicSignups: false # Users need to be invited to have access to Astronomer. Set to true otherwise
      emailConfirmation: true # Users get an email verification before accessing Astronomer
      deployments:
        manualReleaseNames: true # Allows you to set your release names
        serviceAccountAnnotationKey: eks.amazonaws.com/role-arn # Flag to enable using IAM roles (don't enter a specific role)
      email:
        enabled: true
        smtpUrl: YOUR_URI_HERE
        reply: "noreply@astronomer.io" # Emails will be sent from this address
      auth:

>         # Local database (user/pass) configuration.
>         github:
>           enabled: true # Lets users authenticate with Github
>         local:
>           enabled: false # Disables logging in with just a username and password

Reviewer: ```suggestion
        github:
          enabled: true # Lets users authenticate with Github
        local:
          enabled: false # Disables logging in with just a username and password
```

### Astronomer global configuration
#################################
global:

>   # Base domain for all subdomains exposed through ingress

Author: Please confirm that all indentation here is correct.

Reviewer: confirmed

Example wildcard CNAME record:
![aws-elb](https://assets2.astronomer.io/main/docs/ee/route53.png)

## Step 10: Verify You Can Access the Astronomer UI

Go to `app.BASEDOMAIN` to see the Astronomer UI.

Consider this your new Airflow control plane. From the Astronomer UI, you'll be able to both invite and manage users as well as create and monitor Airflow Deployments on the platform.

## Step 11: Verify Your TLS Setup Was Successful

> To check if your TLS certificates were accepted, log into the platform and head to `app.BASEDOMAIN/token` and run:

Reviewer 2: ```suggestion
To check if your TLS certificates were accepted, log in to the Astronomer UI. Then, navigate to `app.BASEDOMAIN/token` and simultaneously run:
```

Example wildcard CNAME record:
![aws-elb](https://assets2.astronomer.io/main/docs/ee/route53.png)

## Step 10: Verify You Can Access the Astronomer UI

Go to `app.BASEDOMAIN` to see the Astronomer UI.

Consider this your new Airflow control plane. From the Astronomer UI, you'll be able to both invite and manage users as well as create and monitor Airflow Deployments on the platform.

## Step 11: Verify Your TLS Setup

To check if your TLS certificates were accepted, log in to the Astronomer UI. Then, go to `app.BASEDOMAIN/token` and run:

```

> curl -v -X POST https://houston.BASEDOMAIN/v1 -H "Authorization: Bearer <token>"

Reviewer 2: ```suggestion
$ curl -v -X POST https://houston.BASEDOMAIN/v1 -H "Authorization: Bearer <token>"
```

```

Verify that this output matches with that of the following command, which doesn't look for TLS:

```

> curl -v -k -X POST https://houston.BASEDOMAIN/v1 -H "Authorization: Bearer <token>"

Reviewer 2: ```suggestion
$ curl -v -k -X POST https://houston.BASEDOMAIN/v1 -H "Authorization: Bearer <token>"
```

```

Next, to make sure the registry is accepted by Astronomer's local docker client, try authenticating to Astronomer with the Astronomer CLI:

```sh
$ astro auth login <your-astronomer-base-domain>
```

If you can log in, then your Docker client trusts the registry. If Docker does not trust the Astronomer registry, run the following and restart Docker:

```
$ mkdir -p /etc/docker/certs.d
$ cp privateCA.pem /etc/docker/certs.d/
```

Finally, try running `$ astro deploy` on a test deployment. Create a deployment in the Astronomer UI, then run:
```sh
$ mkdir demo
$ cd demo
$ astro airflow init
$ astro deploy -f
```
Check the Airflow namespace. If pods are changing at all, then the Houston API trusts the registry.

> If you have Airflow pods in the state "ImagePullBackoff", check the pod description. If you see an x509 error, ensure that you added the `privateCaCertsAddToHost` key-value pairs to your Helm chart. If you missed these during installation, follow the steps in [Apply a Platform Configuration Change on Astronomer](https://www.astronomer.io/docs/enterprise/stable/manage-astronomer/apply-platform-config) to add them after installation.

Reviewer 2: ```suggestion
If you have Airflow pods in an `ImagePullBackoff` state, check the pod description. If you see an x509 error, ensure that you added the `privateCaCertsAddToHost` key-value pairs to your Helm chart. If you missed these during installation, follow the steps in [Apply a Platform Configuration Change on Astronomer](https://www.astronomer.io/docs/enterprise/stable/manage-astronomer/apply-platform-config) to add them after installation.
```

curl -v -k -X POST https://houston.BASEDOMAIN/v1 -H "Authorization: Bearer <token>"
```

> **Note:** The `-k` flag will run the command without looking for TLS

> Finally, to make sure the registry is accepted by Astronomer's local docker client, try logging into Astronomer with the Astronomer CLI:

Reviewer 2: ```suggestion
Finally, to make sure the registry is accepted by Astronomer's local docker client, try authenticating to Astronomer with the Astronomer CLI:
```

```

> astro auth login <your-astronomer-base-domain>

Reviewer 2: ```suggestion
$ astro auth login <your-astronomer-base-domain>
```

> **Note:** The `-k` flag will run the command without looking for SSL

Finally, to make sure the registry is accepted by Astronomer's local docker client, try logging into Astronomer with the Astronomer CLI:

```
astro auth login <your-astronomer-base-domain>
```

If you can log in, then your Docker client trusts the registry. If Docker does not trust the Astronomer registry, run the following and restart Docker:
```
$ mkdir -p /etc/docker/certs.d
$ cp privateCA.pem /etc/docker/certs.d/
```
Next, try running `astro deploy` on a test deployment. Create a deployment in the Astronomer UI, then run:
```
$ mkdir demo
$ cd demo
$ astro airflow init
$ astro deploy -f
```
Check the Airflow namespace. If pods are changing at all, then the Houston API trusts the registry.

> If you have Airflow pods in the state "ImagePullBackoff", check the pod description. If you see an x509 error, ensure that you added the `privateCaCertsAddToHost` key-value pairs to your Helm chart. If you missed these during installation, follow the steps in [Apply a Platform Configuration Change on Astronomer](https://www.astronomer.io/docs/enterprise/stable/manage-astronomer/apply-platform-config) to add them after installation.

Author: This is setup that I inferred based on my general knowledge - please check for accuracy.
